ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Mempool/Init.hs

```diff
@@ -57,27 +57,51 @@ forkSyncStateOnTipPointChange ::
   ResourceRegistry m ->
   MempoolEnv m blk ->
   m ()
-forkSyncStateOnTipPointChange registry menv =
-  void $
+forkSyncStateOnTipPointChange registry menv = do
+  w <-
     forkLinkedWatcher
-      registry
+      (mpEnvRegistry menv)
       "Mempool.syncStateOnTipPointChange"
       Watcher
         { wFingerprint = id
         , wInitial = Nothing
         , wNotify = action
         , wReader = getCurrentTip
         }
+
+  -- With this allocation on the top level registry, we make sure that we first
+  -- stop the watcher thread before closing the mempool registry, as otherwise
+  -- we would run into a race condition (the thread might try to re-sync and
+  -- allocate a forker on the mempool registry which would be closing down).
+  void $ allocate registry (\_ -> pure w) cancelThread
  where
-  action :: Point blk -> m ()
-  action _tipPoint =
-    void $ implSyncWithLedger menv
+  action :: SyncResult m blk -> m ()
+  action (SyncResult a) =
+    void $ implSyncWithLedger menv a
 
   -- Using the tip ('Point') allows for quicker equality checks
-  getCurrentTip :: STM m (Point blk)
+  getCurrentTip :: STM m (SyncResult m blk)
   getCurrentTip =
-    ledgerTipPoint
-      <$> getCurrentLedgerState (mpEnvLedger menv)
+    SyncResult <$> getCurrentLedgerState (mpEnvLedger menv) (mpEnvRegistry menv)
+
+-- | The result of querying the ledger interface for a new state. We use this
+-- newtype only to define the 'Eq' instance such that we compare based on the
+-- point.
+newtype SyncResult m blk
```

Comment:
Would it make sense to use this type elsewhere?

Reply:
I reworked it into `MempoolLedgerDBView`